xlators/features/simple-quota/src/simple-quota.c

```diff
@@ -0,0 +1,1295 @@
+            ret = dict_get_uint64(xdata, GF_GET_FILE_BLOCK_COUNT, &blocks);
+            gf_msg(this->name, GF_LOG_DEBUG, 0, 0,
+                   "reduce size by %" PRIu64 " blocks (ret: %d)", blocks, ret);
+            sq_update_namespace(this, namespace, preparent, postparent,
+                                -((blocks + 1) * 512), "unlink");
+            /* extra 1 for create offset */
+        }
+    }
+
+    frame->local = NULL;
+    inode_unref(namespace);
+    STACK_UNWIND_STRICT(unlink, frame, op_ret, op_errno, preparent, postparent,
+                        xdata);
+    return 0;
+}
+
+int
+sq_unlink(call_frame_t *frame, xlator_t *this, loc_t *loc, int32_t xflag,
+          dict_t *xdata)
+{
+    /* Get the ns inode from parent, it won't cause any changes */
+    xdata = xdata ? dict_ref(xdata) : dict_new();
+    if (!xdata)
+        goto wind;
+    int ret = dict_set_uint32(xdata, GF_REQUEST_LINK_COUNT_XDATA, 1);
+    if (IS_ERROR(ret))
+        gf_log(this->name, GF_LOG_ERROR,
+               "dict set failed (pargfid: %s, name: %s), "
+               "still continuing",
+               uuid_utoa(loc->pargfid), loc->name);
+
+    ret = dict_set_uint64(xdata, GF_GET_FILE_BLOCK_COUNT, 1);
+    if (IS_ERROR(ret))
+        gf_log(this->name, GF_LOG_ERROR,
+               "dict set failed (pargfid: %s, name: %s), "
+               "still continuing",
+               uuid_utoa(loc->pargfid), loc->name);
+
+wind:
+    frame->local = inode_ref(loc->parent->ns_inode);
+    STACK_WIND(frame, sq_unlink_cbk, FIRST_CHILD(this),
+               FIRST_CHILD(this)->fops->unlink, loc, xflag, xdata);
+    if (xdata)
+        dict_unref(xdata);
+    return 0;
+}
+
+int32_t
+sq_rmdir_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int32_t op_ret,
+             int32_t op_errno, struct iatt *preparent, struct iatt *postparent,
+             dict_t *xdata)
+{
+    inode_t *namespace = frame->local;
+
+    if (IS_SUCCESS(op_ret)) {
+        /* Just remove 1 4k block */
+        // sq_update_namespace(this, namespace, preparent, postparent, -4096,
+        // "rmdir");
+    }
+
+    frame->local = NULL;
+    inode_unref(namespace);
+    STACK_UNWIND_STRICT(rmdir, frame, op_ret, op_errno, preparent, postparent,
+                        xdata);
+    return 0;
+}
+
+int
+sq_rmdir(call_frame_t *frame, xlator_t *this, loc_t *loc, int flag,
+         dict_t *xdata)
+{
+    frame->local = inode_ref(loc->parent->ns_inode);
+    STACK_WIND(frame, sq_rmdir_cbk, FIRST_CHILD(this),
+               FIRST_CHILD(this)->fops->rmdir, loc, flag, xdata);
+    return 0;
+}
+
+int32_t
+sq_create_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int32_t op_ret,
+              int32_t op_errno, fd_t *fd, inode_t *inode, struct iatt *buf,
+              struct iatt *preparent, struct iatt *postparent, dict_t *xdata)
+{
+    inode_t *namespace = frame->local;
+
+    if (IS_SUCCESS(op_ret)) {
+        sq_update_namespace(this, namespace, preparent, postparent, 512,
```

Comment:
Should this be ia_blksize instead?

Reply:
taking it as 'ia_blocks' so, expecting it to be 512.

Reply:
> taking it as 'ia_blocks' so, expecting it to be 512.
hi , I don't understand why you need to add 512 when creating a file? When the number of files is large, it will be found that df is much larger than du statistics

Reply:
I added the counter in `create()` and `mkdir()` mainly to prevent someone abusing the whole filesystem by creating lots of files, and no content in it. Specially in the case of containers, with simple quota, you can set quota limit as minimum as few MBs, and if I don't add a 'cost' to file creates the filesystem may get abused. Also note, for every few files created, the size of the 'directory' also grows, which is reflected in the stat information (but is not captured as a 'write' event), thus, adding this 512 also meets that extra counts.